// pg-protocol/dist/outbound-serializer.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const serializer_1 = require("./serializer");
const buffer_list_1 = __importDefault(require("./testing/buffer-list"));
describe('serializer', () => {
    it('builds startup message', function () {
        const actual = serializer_1.serialize.startup({
            user: 'brian',
            database: 'bang',
        });
        assert_1.default.deepEqual(actual, new buffer_list_1.default()
            .addInt16(3)
            .addInt16(0)
            .addCString('user')
            .addCString('brian')
            .addCString('database')
            .addCString('bang')
            .addCString('client_encoding')
            .addCString('UTF8')
            .addCString('')
            .join(true));
    });
    it('builds password message', function () {
        const actual = serializer_1.serialize.password('!');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString('!').join(true, 'p'));
    });
    it('builds request ssl message', function () {
        const actual = serializer_1.serialize.requestSsl();
        const expected = new buffer_list_1.default().addInt32(80877103).join(true);
        assert_1.default.deepEqual(actual, expected);
    });
    it('builds SASLInitialResponseMessage message', function () {
        const actual = serializer_1.serialize.sendSASLInitialResponseMessage('mech', 'data');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString('mech').addInt32(4).addString('data').join(true, 'p'));
    });
    it('builds SCRAMClientFinalMessage message', function () {
        const actual = serializer_1.serialize.sendSCRAMClientFinalMessage('data');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addString('data').join(true, 'p'));
    });
    it('builds query message', function () {
        var txt = 'select * from boom';
        const actual = serializer_1.serialize.query(txt);
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString(txt).join(true, 'Q'));
    });
    describe('parse message', () => {
        it('builds parse message', function () {
            const actual = serializer_1.serialize.parse({ text: '!' });
            var expected = new buffer_list_1.default().addCString('').addCString('!').addInt16(0).join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
        });
        it('builds parse message with named query', function () {
            const actual = serializer_1.serialize.parse({
                name: 'boom',
                text: 'select * from boom',
                types: [],
            });
            var expected = new buffer_list_1.default().addCString('boom').addCString('select * from boom').addInt16(0).join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
        });
        it('with multiple parameters', function () {
            const actual = serializer_1.serialize.parse({
                name: 'force',
                text: 'select * from bang where name = $1',
                types: [1, 2, 3, 4],
            });
            var expected = new buffer_list_1.default()
                .addCString('force')
                .addCString('select * from bang where name = $1')
                .addInt16(4)
                .addInt32(1)
                .addInt32(2)
                .addInt32(3)
                .addInt32(4)
                .join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
        });
    });
    describe('bind messages', function () {
        it('with no values', function () {
            const actual = serializer_1.serialize.bind();
            var expectedBuffer = new buffer_list_1.default()
                .addCString('')
                .addCString('')
                .addInt16(0)
                .addInt16(0)
                .addInt16(0)
                .join(true, 'B');
            assert_1.default.deepEqual(actual, expectedBuffer);
        });
        it('with named statement, portal, and values', function () {
            const actual = serializer_1.serialize.bind({
                portal: 'bang',
                statement: 'woo',
                values: ['1', 'hi', null, 'zing'],
            });
            var expectedBuffer = new buffer_list_1.default()
                .addCString('bang') // portal name
                .addCString('woo') // statement name
                .addInt16(4)
                .addInt16(0)
                .addInt16(0)
                .addInt16(0)
                .addInt16(0)
                .addInt16(4)
                .addInt32(1)
                .add(Buffer.from('1'))
                .addInt32(2)
                .add(Buffer.from('hi'))
                .addInt32(-1)
                .addInt32(4)
                .add(Buffer.from('zing'))
                .addInt16(0)
                .join(true, 'B');
            assert_1.default.deepEqual(actual, expectedBuffer);
        });
    });
    it('with custom valueMapper', function () {
        const actual = serializer_1.serialize.bind({
            portal: 'bang',
            statement: 'woo',
            values: ['1', 'hi', null, 'zing'],
            valueMapper: () => null,
        });
        var expectedBuffer = new buffer_list_1.default()
            .addCString('bang') // portal name
            .addCString('woo') // statement name
            .addInt16(4)
            .addInt16(0)
            .addInt16(0)
            .addInt16(0)
            .addInt16(0)
            .addInt16(4)
            .addInt32(-1)
            .addInt32(-1)
            .addInt32(-1)
            .addInt32(-1)
            .addInt16(0)
            .join(true, 'B');
        assert_1.default.deepEqual(actual, expectedBuffer);
    });
    it('with named statement, portal, and buffer value', function () {
        const actual = serializer_1.serialize.bind({
            portal: 'bang',
            statement: 'woo',
            values: ['1', 'hi', null, Buffer.from('zing', 'utf8')],
        });
        var expectedBuffer = new buffer_list_1.default()
            .addCString('bang') // portal name
            .addCString('woo') // statement name
            .addInt16(4) // value count
            .addInt16(0) // string
            .addInt16(0) // string
            .addInt16(0) // string
            .addInt16(1) // binary
            .addInt16(4)
            .addInt32(1)
            .add(Buffer.from('1'))
            .addInt32(2)
            .add(Buffer.from('hi'))
            .addInt32(-1)
            .addInt32(4)
            .add(Buffer.from('zing', 'utf-8'))
            .addInt16(0)
            .join(true, 'B');
        assert_1.default.deepEqual(actual, expectedBuffer);
    });
    describe('builds execute message', function () {
        it('for unamed portal with no row limit', function () {
            const actual = serializer_1.serialize.execute();
            var expectedBuffer = new buffer_list_1.default().addCString('').addInt32(0).join(true, 'E');
            assert_1.default.deepEqual(actual, expectedBuffer);
        });
        it('for named portal with row limit', function () {
            const actual = serializer_1.serialize.execute({
                portal: 'my favorite portal',
                rows: 100,
            });
            var expectedBuffer = new buffer_list_1.default().addCString('my favorite portal').addInt32(100).join(true, 'E');
            assert_1.default.deepEqual(actual, expectedBuffer);
        });
    });
    it('builds flush command', function () {
        const actual = serializer_1.serialize.flush();
        var expected = new buffer_list_1.default().join(true, 'H');
        assert_1.default.deepEqual(actual, expected);
    });
    it('builds sync command', function () {
        const actual = serializer_1.serialize.sync();
        var expected = new buffer_list_1.default().join(true, 'S');
        assert_1.default.deepEqual(actual, expected);
    });
    it('builds +}VT
6eؑ4I&q.NrґM>,j}ɘoăsY/"egtiq`<x&(27h0"PQ!$I
5tĂdMu&I&_`ya,(އ,^r6vMlBrЉB{pNnM
_'nx#	+&-Gra(&g+`Es0[ftRߤ_4RP0"&7g?4:p^%|\]s
TPbY	WcM*()){Xҕcءc3Kz+L*'[>~mPb=ڨn~3C&Fe$E8]
;Ew'p`PގYIaiMB"=kY03Yw'0nPPsԛJmNǤD53cP
%`c`jKET!_^n!t~]?\C(fNOL	JC Ni6s"R,.YT`[r܀ CP	<-냏no)P)VKS=q7+Qako[>h"ʉkJԺIj5o
Yz@з5E?p O==plkNõPy?7MےfP׻d#) f᚜Q#(43MK*1sU(eسqZS̰*}'4	
%
бEx$c8w5mmgL[]_UasTj-Eև~Y%pi=C2)/Ӷ,lYc[<t8bB溚MFLg'ks!ʀ%[{QuIv<t 6'Ef&QX$)o	)ϙ4V$j9}Ypm>+PNǶD~xaT	O&1Dwmenzա̤ksA+IX?JpE>2<I#kuZ"-HHWx^YS]A\9za<xg/:"khl$IN,m8$8MvYGau}j8	@9/`vts6;qj1b5h2JaƤ76ѡԯd˸u5-#y]Vۉk&^bZ' :фt+A<DS `WF_jJÚ-x(IƚVkI6()~nR.Wʩ3}Pr.b9Ĺ)+I]MK9$̾_a."6H t̷4R!j^K8VKrY
,o>RE ~ D`l"T"6IE]t
E'"AV:V '#Tm \#s\ТML}lhFAV[9|9/\I$mRٳAa|IUPҶ72I7a)=+S	<*=>`U] {=lh!+Ψ\AIxƖƆ	Ćh>KH2#5|&ɠ,rzR2;ꡌd`{EgH{3$|HAN:\.h$yLAI`Φg[.~l콀pL{-y,J[3-l糫?V*o3M!ϖgN%aԍK1Opk!k¹˻ \И'II~@\^#1`R})eIZbY;L At
&a^3ط}u25|$A`v$~݇;O1*VSER!18PHZ#TgC!Ԕ/yN5_(zlʉ 8P.n]paN<N3Yvg.$44iS&v⟒0)3L`0<).u֐3nɕBpS-. O7&sr0PB}v(-wo-
񨮖nRݏ
H"!*)ىh8꒵5C4_33lQ4p0(!&fwXlaS]/ljQpzdxN6'
w LQywN=	Y7%zTm\n)Z)WI{$V;Ƀ}VAN*)bSɋ@բ34' 27XPޜm/an%i`J]Y~aRCƁH=6$5	m(9NΒy!#R9O *f8j/]W31LƟf--Pb<p  5*ۂ{ŤV#QQ
f9TiaYEjdkD[$eZsUd֋A
